File: julia/words/src/Words.jl
#!/usr/bin/julia

module Words
export reverse_string, is_palindrome, kmp_search, lcs

function reverse_string(str::String) :: String
  io = IOBuffer(write=true)
  for i = length(str):-1:1
    write(io, str[i])
    i -= 1
  end
  return String(take!(io))
end

function is_palindrome(str::String) :: Bool
  cleaned_str = replace(str, r"\s+" => "")
  i = 1
  j = length(cleaned_str)
  while i < j
    if cleaned_str[i] != cleaned_str[j]
      return false
    end
    i += 1
    j -= 1
  end
  return true
end

function compute_lps_array(pat::String) :: Array{Int,1}
  len_pat = length(pat)
  lps = Array{Int, 1}(undef, len_pat)
  lps[1] = 0
  i = 2
  len = 0
  while i <= len_pat
    if pat[i] == pat[len + 1]
      len += 1
      lps[i] = len
      i += 1
    else
      if len > 0
        len = lps[len]
      else
        lps[i] = 0
        i += 1
      end
    end
  end
  return lps
end

function kmp_search(sub::String, str::String) :: Bool
  len_sub = length(sub)
  len_str = length(str)
  if len_sub == 0
    return true
  elseif len_sub > len_str
    return false
  end
  lps = compute_lps_array(sub)
  i = 1
  j = 1
  while i <= len_str
    if str[i] == sub[j]
      i += 1
      j += 1
      if j > len_sub
        return true
      end
    else
      if j > 1
        j = lps[j - 1]
        if j == 0
          j += 1
        end
      else
        i += 1
      end
    end
  end
  return false
end

function create_matrix(rows::Int, cols::Int)::Array{Int, 2}
  arr = Array{Int, 2}(undef, rows, cols)
  for i in 1:rows
    for j in 1:cols
      arr[i, j] = 0
    end
  end
  return arr
end

function lcs(str1::String, str2::String) :: Int
  len1 = length(str1)
  len2 = length(str2)
  dp = create_matrix(len1 + 1, len2 + 1)
  for i = 2:(len1 + 1)
    for j = 2:(len2 + 1)
      if str1[i - 1] == str2[j - 1]
        dp[i, j] = dp[i - 1, j - 1] + 1
      else
        dp[i, j] = max(dp[i - 1, j], dp[i, j - 1])
      end
    end
  end
  return dp[len1 + 1, len2 + 1]
end

end

File: nim/words/src/words.nim
# This is just an example to get you started. A typical library package
# exports the main API in this file. Note that you cannot rename this file
# but you can remove it if you wish.
import re

proc add*(x, y: int): int =
  ## Adds two numbers together.
  return x + y

proc reverseString*(str: string): string =
  var reversed = ""
  for i in countdown(len(str) - 1, 0):
    reversed.add(str[i])
  return reversed

proc isPalindrome*(str: string): bool =
  var cleanedStr = str.replace(re"\s+", "")
  var i = 0
  var j = len(cleanedStr) - 1
  while i < j:
    if cleanedStr[i] != cleanedStr[j]:
      return false
    i += 1
    j -= 1
  return true

proc computeLpsArray*(patt: string): seq[int] =
  let lenPatt = len(patt)
  var lps: seq[int]
  newSeq(lps, lenPatt)
  lps[0] = 0
  var i = 1
  var len = 0
  while i < lenPatt:
    if patt[i] == patt[len]:
      len += 1
      lps[i] = len
      i += 1
    else:
      if len != 0:
        len = lps[len - 1]
      else:
        lps[i] = 0
        i += 1
  return lps

proc kmpSearch*(sub, str: string): bool =
  let lenSub = len(sub)
  let lenStr = len(str)
  if lenSub == 0:
    return true
  elif lenSub > lenStr:
    return false
  let lps = computeLpsArray(sub)
  var i = 0
  var j = 0
  while i < lenStr:
    if str[i] == sub[j]:
      i += 1
      j += 1
      if j == lenSub:
        return true
    else:
      if j != 0:
        j = lps[j - 1]
      else:
        i += 1
  return false

proc lcs*(str1, str2: string): int =
  let len1 = len(str1)
  let len2 = len(str2)
  var dp: seq[seq[int]]
  newSeq(dp, len1 + 1)
  newSeq(dp[0], len2 + 1)
  for i in 1..len1:
    newSeq(dp[i], len2 + 1)
    for j in 1..len2:
      if str1[i - 1] == str2[j - 1]:
        dp[i][j] = dp[i - 1][j - 1] + 1
      else:
        dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])
  return dp[len1][len2]